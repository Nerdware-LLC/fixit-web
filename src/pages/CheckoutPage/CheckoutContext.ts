import { createContext, useContext } from "react";
import type { EncodedAuthToken, UserSubscriptionPriceLabel } from "../../types";

export interface CheckoutContextValues {
  selectedSubscription: UserSubscriptionPriceLabel;
  promoCode: string | null;
  handlePostSubmit: ({ token }: { token: EncodedAuthToken }) => Promise<void>;
}

export const CheckoutContext = createContext<CheckoutContextValues>({
  selectedSubscription: "TRIAL",
  promoCode: null,
  handlePostSubmit: async ({ token }: { token: EncodedAuthToken }) => {}
});

export const useCheckoutContext = () => useContext(CheckoutContext);
