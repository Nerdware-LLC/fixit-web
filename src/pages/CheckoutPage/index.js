export { CheckoutPage as default } from "./CheckoutPage";
// export as default to enable code-splitting via lazy import.
