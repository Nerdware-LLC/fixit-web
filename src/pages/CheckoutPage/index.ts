export { Component } from "./CheckoutPage";
// Exported as "Component" for react-router-dom lazy loading
