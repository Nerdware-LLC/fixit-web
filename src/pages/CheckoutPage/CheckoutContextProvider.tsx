import React, { useState, useEffect } from "react";
import { useLocation, useNavigate } from "react-router-dom";
import { CheckoutContext, type CheckoutContextValues } from "./CheckoutContext";
import { useWebViewContext, Loading } from "../../components";
import { CONSTANTS } from "../../types";
import { useQueryParams, useAuthToken } from "../../hooks";
import { storage, typeSafePropertyExists } from "../../utils";

export const CheckoutContextProvider = ({ children }: { children: React.ReactNode }) => {
  const [checkoutValues, setCheckoutValues] = useState<CheckoutContextValues>();

  // For mobile context:
  const { isAppWithinWebView, webViewPostMessage } = useWebViewContext();
  const queryParamsFromMobile = useQueryParams();
  // For non-mobile context:
  const { state: locationState } = useLocation();
  const { processAuthToken } = useAuthToken();
  const nav = useNavigate();

  /* Checkout values will either come from queryParams if used within a
  mobile webview, or location.state if used in a non-mobile context.

  Once CheckoutForm submits the payment info, it passes the updated authToken
  into the `handlePostSubmit` fn. If this is operating within a mobile webview,
  that token is sent back to the mobile app via the `webViewPostMessage` fn; if
  not, the token is submitted for processing so the user's auth/payment info can
  be updated in the local cache.  */

  useEffect(() => {
    if (isAppWithinWebView) {
      // IF MOBILE:
      if (!areInitialCheckoutContextValuesValid(queryParamsFromMobile)) {
        webViewPostMessage({ error: "INVALID_QUERY_PARAMS" });
      } else {
        storage.setAuthToken(queryParamsFromMobile.token);
        setCheckoutValues({
          selectedSubscription: queryParamsFromMobile.sub.toUpperCase(),
          promoCode: queryParamsFromMobile?.promoCode ?? null,
          handlePostSubmit: webViewPostMessage
        } as CheckoutContextValues);
      }
    } else {
      // ELSE NON-MOBILE:
      if (!locationState || !areInitialCheckoutContextValuesValid(locationState)) {
        nav("/products", { replace: true, state: { isRedirect: true } });
      } else {
        setCheckoutValues({
          selectedSubscription: locationState.sub.toUpperCase(),
          promoCode: locationState?.promoCode ?? null,
          handlePostSubmit: ({ token: updatedAuthToken }) => {
            processAuthToken(updatedAuthToken).then(() => nav("/home"));
          }
        } as CheckoutContextValues);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return !checkoutValues ? (
    <Loading />
  ) : (
    <CheckoutContext.Provider value={checkoutValues}>{children}</CheckoutContext.Provider>
  );
};

const areInitialCheckoutContextValuesValid = (obj: Record<string, any>) => {
  // prettier-ignore
  return (
    typeSafePropertyExists(obj, "sub")
    && CONSTANTS.USER_SUBSCRIPTION.PRICE_LABELS.includes(obj.sub)
    && typeSafePropertyExists(obj, "promoCode")
    && (typeof obj.promoCode === "string" || obj.promoCode === null)
  );
};
