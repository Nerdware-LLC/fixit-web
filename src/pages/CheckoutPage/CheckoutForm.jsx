import React from "react";
import { toast } from "react-toastify";
import styled from "@emotion/styled";
import { useCheckoutContext } from "./CheckoutContext";
import { StripeForm } from "../../components";
import { stripeService } from "../../services";

export const CheckoutForm = () => {
  const { selectedSubscription, promoCode, handlePostSubmit } = useCheckoutContext();

  const handleSubmit = async (paymentMethod) => {
    const { token } = await stripeService.submitPaymentForSubscription({
      selectedSubscription,
      promoCode,
      paymentMethodID: paymentMethod.id
    });
    toast.success("Success! 👍 You will be redirected in just a moment.", {
      onClose: () => handlePostSubmit({ token })
    });
  };

  return <StripeCheckoutForm handleSubmit={handleSubmit} />;
};

const StripeCheckoutForm = styled(StripeForm)`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
`;
