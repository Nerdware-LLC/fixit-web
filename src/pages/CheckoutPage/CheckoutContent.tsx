import { styled } from "@mui/material/styles";
import Divider from "@mui/material/Divider";
import Paper from "@mui/material/Paper";
import Text from "@mui/material/Typography";
import { FetchStateContextProvider } from "@/app/FetchStateContext";
import { usePageLayoutContext } from "@/app/PageLayoutContext/usePageLayoutContext";
import { TitleLogo, brandingClassNames } from "@/components/Branding";
import { DemoStripeCardInfoAccordion } from "@/components/DevTools/DemoStripeCardInfoAccordion";
import { CheckoutForm, type CheckoutFormProps } from "./CheckoutForm";
import { PaymentConfirmationInfo } from "./PaymentConfirmationInfo";
import { SubCostDetails } from "./SubCostDetails";
import { checkoutPageClassNames } from "./classNames";

export const CheckoutContent = ({
  showPaymentConfirmation,
  onCheckoutCompletion,
}: CheckoutContentProps) => {
  const { isMobilePageLayout } = usePageLayoutContext();

  return (
    <StyledPaper elevation={0}>
      {/*   SECTION: LEFT/TOP   */}

      <div className={checkoutPageClassNames.pageSectionColumn}>
        <TitleLogo
          sx={{
            [`& .${brandingClassNames.titleLogoImg}`]: { height: "8vh" },
            [`& .${brandingClassNames.titleLogoText}`]: { fontSize: "3rem" },
          }}
        />
        <SubCostDetails />
      </div>

      {isMobilePageLayout && <Divider />}

      {/*   SECTION: RIGHT/BOTTOM   */}

      <div className={checkoutPageClassNames.pageSectionColumn}>
        {showPaymentConfirmation ? (
          <PaymentConfirmationInfo />
        ) : (
          <>
            <Text style={{ textAlign: "center", opacity: 0.75, whiteSpace: "pre-line" }}>
              {`You'll receive a confirmation email\nwith receipt upon completion.`}
            </Text>
            <DemoStripeCardInfoAccordion />
            <Divider />
            <FetchStateContextProvider>
              <CheckoutForm onSuccessfulSubmit={onCheckoutCompletion} />
            </FetchStateContextProvider>
          </>
        )}
      </div>
    </StyledPaper>
  );
};

const StyledPaper = styled(Paper)(({ theme: { variables } }) => ({
  width: "100%",
  minHeight: "600px",
  textAlign: "left",
  border: "1px solid rgba(255, 255, 255, 0.35)",
  borderRadius: "10px",
  display: "flex",

  ...(variables.isMobilePageLayout
    ? {
        maxWidth: "480px", // 30rem, any larger looks too wide for the content on mobile
        flexDirection: "column",
        padding: "1rem",
        gap: "1rem",
      }
    : {
        maxWidth: "1040px", // 65rem, any larger looks too wide for the content on desktop
        flexDirection: "row",
        justifyContent: "space-between",
        padding: "2rem",
        gap: "2rem",
      }),

  [`& > div.${checkoutPageClassNames.pageSectionColumn}`]: {
    ...(variables.isMobilePageLayout
      ? {
          width: "100%",
          height: "min-content",
        }
      : {
          width: "50%",
          minHeight: "100%",
        }),
    gap: "inherit",
    display: "flex",
    flexDirection: "column",
    justifyContent: "space-evenly",
  },
}));

export type CheckoutContentProps = {
  showPaymentConfirmation: boolean;
  onCheckoutCompletion: CheckoutFormProps["onSuccessfulSubmit"];
};
