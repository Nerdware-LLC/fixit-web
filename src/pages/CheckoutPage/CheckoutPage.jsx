import React, { useEffect } from "react";
import styled from "@emotion/styled";
import { CheckoutContext } from "./CheckoutContext";
import { HeaderIcon } from "./HeaderIcon";
import { ExplainerText } from "./ExplainerText";
import { SubCostDetails } from "./SubCostDetails";
import { CheckoutForm } from "./CheckoutForm";
import { CheckoutStripeBadge } from "./CheckoutStripeBadge";
import { useWebViewContext } from "../../components";
import { storage, useQueryParams } from "../../utils";

export const CheckoutPage = () => {
  const { webViewPostMessage } = useWebViewContext();
  const { token, sub, promoCode } = useQueryParams();

  useEffect(() => {
    storage.setAuthToken(token);
  }, [token]);

  if (!token || !sub) webViewPostMessage({ error: "INVALID_QUERY_PARAMS" });

  return (
    <CheckoutContext.Provider
      value={{
        selectedSubscription: sub.toUpperCase(),
        promoCode,
        handlePostSubmit: webViewPostMessage
      }}
    >
      <StyledContentContainer>
        <HeaderIcon />
        <ExplainerText />
        <SubCostDetails />
        <CheckoutForm />
        <CheckoutStripeBadge />
      </StyledContentContainer>
    </CheckoutContext.Provider>
  );
};

const StyledContentContainer = styled.div`
  position: absolute;
  min-height: 80vh;
  width: 85vw;
  padding: 3.5rem 1.75rem 1.75rem 1.75rem;
  text-align: left;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  border-radius: 10px;
  background-color: ${({ theme }) => theme.palette.background.paper};
  border: 1px solid rgba(255, 255, 255, 0.35);
`;
