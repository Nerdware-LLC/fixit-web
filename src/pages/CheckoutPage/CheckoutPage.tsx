import { useState, useEffect } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";
import dayjs from "dayjs";
import { styled } from "@mui/material/styles";
import Backdrop from "@mui/material/Backdrop";
import Box from "@mui/material/Box";
import Divider from "@mui/material/Divider";
import Paper, { paperClasses } from "@mui/material/Paper";
import Text, { typographyClasses } from "@mui/material/Typography";
import { usePageLayoutContext } from "@app/PageLayoutContext/usePageLayoutContext";
import { useWebViewContext } from "@app/WebViewContext/useWebViewContext";
import { checkoutValuesStore } from "@cache/checkoutValuesStore";
import { TitleLogo, titleLogoClassNames } from "@components/Branding/TitleLogo";
import { FetchStateContextWrapper } from "@components/Indicators/FetchStateContextWrapper";
import { Loading } from "@components/Indicators/Loading";
import { useLottie } from "@components/LottieAnimations/useLottie";
import { LegalLinks } from "@components/Navigation/LegalLinks";
import { storage } from "@utils/storage";
import {
  USER_SUBSCRIPTION_PRICE_LABELS,
  type UserSubscriptionPriceLabel,
} from "@/types/UserSubscription";
import { CheckoutForm } from "./CheckoutForm";
import { PaymentConfirmationInfo } from "./PaymentConfirmationInfo";
import { SubCostDetails } from "./SubCostDetails";
import { checkoutPageClassNames } from "./classNames";

export const CheckoutPage = () => {
  const { selectedSubscription } = checkoutValuesStore.useSubToStore();
  const nav = useNavigate();
  const { isMobilePageLayout } = usePageLayoutContext();

  // For mobile-app WebView:
  const { isAppWithinWebView, webViewPostMessage } = useWebViewContext();
  const [searchParams] = useSearchParams();

  useEffect(() => {
    if (isAppWithinWebView) {
      // If in a mobile-app WebView, "token" and "sub" query params must exist
      const token = searchParams.get("token");
      const sub = searchParams.get("sub");
      const promoCode = searchParams.get("promoCode");

      if (!token || !sub || !USER_SUBSCRIPTION_PRICE_LABELS.includes(sub as any)) {
        webViewPostMessage({ error: "INVALID_QUERY_PARAMS" });
      } else {
        storage.authToken.set(token);
        checkoutValuesStore.set({
          selectedSubscription: sub as UserSubscriptionPriceLabel,
          promoCode,
        });
      }
    } else if (!USER_SUBSCRIPTION_PRICE_LABELS.includes(selectedSubscription as any)) {
      // Else if not in a mobile-app WebView, redirect to /products if "selectedSubscription" isn't set
      nav("/products", { replace: true, state: { isRedirect: true } });
    }
  }, [isAppWithinWebView, nav, searchParams, selectedSubscription, webViewPostMessage]);

  // For displaying payment confirmation:
  const { LottieView, playLottie } = useLottie({ animation: "success-checkmark" });
  const [showBackdrop, setShowBackdrop] = useState<boolean>(false);
  const [showPaymentConfirmation, setShowPaymentConfirmation] = useState<boolean>(false);

  const handleCompleteCheckout = async (success: boolean) => {
    if (success) {
      setShowBackdrop(true);
      await playLottie();
      setShowBackdrop(false);
      setShowPaymentConfirmation(true);
    }
  };

  return !selectedSubscription ? (
    <Loading />
  ) : (
    <FetchStateContextWrapper>
      <StyledBox>
        <Paper elevation={0}>
          {/*   SECTION: LEFT/TOP   */}

          <div className={checkoutPageClassNames.pageSectionColumn}>
            <TitleLogo />
            <div>
              <SubCostDetails />
            </div>
          </div>

          {/*   SECTION: RIGHT/BOTTOM   */}

          <div className={checkoutPageClassNames.pageSectionColumn}>
            {isMobilePageLayout && <Divider />}
            <Box className={checkoutPageClassNames.paymentInfoTextBox}>
              {
                // Show additional info on larger displays:
                !showPaymentConfirmation && !isMobilePageLayout && (
                  <>
                    <Text>
                      People who need to get things done use Fixit to keep in touch with contractors
                      and customers, create work orders, submit invoices, and manage payments - all
                      in one place.
                    </Text>
                    <br />
                    <Text>
                      Whether you&apos;re a homeowner planning your next kitchen renovation, or a
                      general contractor looking for a better way to submit invoices and get paid
                      for your work, Fixit makes it easy.
                    </Text>
                    <br />
                  </>
                )
              }
              {showPaymentConfirmation !== true && (
                <Text className={checkoutPageClassNames.paymentConfirmationText}>
                  You'll receive a confirmation email with receipt upon completion.
                </Text>
              )}
            </Box>
            {showPaymentConfirmation !== true ? (
              <>
                <Text variant="h2" className={checkoutPageClassNames.checkoutFormHeader}>
                  {selectedSubscription === "TRIAL" ? "Enter payment details" : "Pay with card"}
                </Text>
                <CheckoutForm onCompleteCheckout={handleCompleteCheckout} />
                <Text className={checkoutPageClassNames.checkoutFormLegalText}>
                  {
                    // The below dayjs format string "LL" results in dates like "January 1, 2020"
                    // prettier-ignore
                    selectedSubscription === "TRIAL"
                        ? `After your trial ends, you will be charged $5.00 per month starting ${dayjs().add(14, "days").format("LL")}. You can always cancel before then.`
                        : `By confirming your subscription, you allow Fixit to charge your card for this payment and future payments in accordance with their terms. You can always cancel your subscription.`
                  }
                </Text>
              </>
            ) : (
              <PaymentConfirmationInfo />
            )}
          </div>
        </Paper>
        <LegalLinks includeStripeBadge style={{ margin: "1.5rem" }} />
        <Backdrop open={showBackdrop}>{LottieView}</Backdrop>
      </StyledBox>
    </FetchStateContextWrapper>
  );
};

const StyledBox = styled(Box)(({ theme }) => ({
  minHeight: "100%",
  width: "100%",
  padding: "clamp(1rem, 5%, 2rem)",
  paddingBottom: 0,
  display: "flex",
  flexDirection: "column",
  justifyContent: "space-between",
  alignItems: "center",
  "@media (min-width: 1450px)": {
    padding: "2rem clamp(2rem, 15%, 25rem) 0 clamp(2rem, 15%, 25rem)",
  },

  // BASE TEXT STYLE (from the old <StyledText>)
  [`& .${checkoutPageClassNames.baseText}`]: {
    fontSize: "clamp(1rem, 4vw, 1.15rem)",
    lineHeight: "1.5rem",
  },

  [`& > .${paperClasses.root}`]: {
    flexGrow: 1,
    height: "100%",
    width: "100%",
    textAlign: "left",
    border: "1px solid rgba(255, 255, 255, 0.35)",
    borderRadius: "10px",
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    ...(theme.variables.isMobilePageLayout
      ? { padding: "1rem", flexWrap: "wrap" }
      : { padding: "2rem", flexWrap: "nowrap" }),

    [`& > div.${checkoutPageClassNames.pageSectionColumn}`]: {
      width: theme.variables.isMobilePageLayout ? "100%" : "50%",
      minHeight: "100%",
      display: "flex",
      flexDirection: "column",
      justifyContent: "space-between",

      // SECTION: LEFT/TOP
      "&:first-of-type": {
        [`& > .${titleLogoClassNames.root}`]: {
          justifyContent: "flex-start",
          marginBottom: "1rem",
          [`& > .${titleLogoClassNames.logoImg}`]: {
            height: "8vh",
            marginRight: "0.5rem",
          },
          [`& > .${titleLogoClassNames.logoText}`]: {
            fontSize: "3rem",
          },
        },

        // SubCostDetails container // TODO should be able to rm this div layer
        "& > div:nth-of-type(2)": {
          height: "100%",
          ...(!theme.variables.isMobilePageLayout && { padding: "2.5rem 0 1.5rem 0" }),
        },
      },

      // SECTION: RIGHT/BOTTOM
      "&:nth-of-type(2)": {
        ...(!theme.variables.isMobilePageLayout && {
          padding: "2rem 0 2rem 2rem",
        }),

        [`& > .${checkoutPageClassNames.paymentInfoTextBox}`]: {
          width: "100%",
          opacity: "0.75",
          [`& .${typographyClasses.root}`]: {
            fontSize: "1.15rem",
            lineHeight: "1.5rem",
          },

          [`& > .${checkoutPageClassNames.paymentConfirmationText}`]: {
            ...(theme.variables.isMobilePageLayout && {
              margin: "1rem 0",
              textAlign: "center",
              fontSize: "0.95rem",
              lineHeight: "1.15rem",
              fontWeight: 300,
            }),
          },
        },

        [`& > .${checkoutPageClassNames.checkoutFormHeader}`]: {
          margin: theme.variables.isMobilePageLayout ? "1rem 0" : "3rem 0",
          padding: 0,
          fontSize: "2rem",
          fontWeight: 400,
          borderWidth: "0 0 1px 0",
          borderStyle: "solid",
          borderColor: theme.palette.divider,
        },

        [`& > .${checkoutPageClassNames.checkoutFormLegalText}`]: {
          marginTop: "0.5rem",
          textAlign: "center",
          opacity: 0.75,
          fontSize: "0.7rem",
          lineHeight: "1rem",
          fontWeight: 100,
        },
      },
    },
  },
}));
