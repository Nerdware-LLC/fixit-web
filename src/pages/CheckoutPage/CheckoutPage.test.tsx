import renderer from "react-test-renderer";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { CheckoutPage } from "./CheckoutPage";
import { ThemeProvider } from "@app/ThemeProvider";

it("renders correctly", () => {
  expect(
    renderer
      .create(
        <ThemeProvider>
          <MemoryRouter
            initialEntries={[
              "/checkout?token=TEST_TOKEN_VALUE&sub=annual&promoCode=TEST_PROMO_CODE"
            ]}
          >
            <Routes>
              <Route path="/checkout" element={<CheckoutPage />} />
            </Routes>
          </MemoryRouter>
        </ThemeProvider>
      )
      .toJSON()
  ).toMatchSnapshot();
});
