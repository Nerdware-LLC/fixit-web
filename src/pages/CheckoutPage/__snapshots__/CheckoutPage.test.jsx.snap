// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="css-nk4bsi"
>
  <img
    className="css-tg2ra5"
    src="fixit-icon.png"
  />
  <div
    className="css-dcigc0"
  >
    <p
      className="css-1a9s9ng"
    >
      Enter your card details.
    </p>
    <p
      className="css-1a9s9ng"
      style={
        Object {
          "marginTop": "0.5rem",
        }
      }
    >
      Your subscription will begin as soon as 
      your payment is submitted.
    </p>
    <ul
      className="css-gxgq0b"
    >
      <li>
        <span>
          Billing will occur automatically around the same day each year.
        </span>
      </li>
      <li
        style={
          Object {
            "margin": "0.5rem 0",
          }
        }
      >
        <span>
          You can change or cancel your subscription at any time.
        </span>
      </li>
      <li>
        <span>
          You'll receive a confirmation email
           
          with receipt 
          upon completion, as well as regular emails regarding account updates and other Fixit info.
        </span>
      </li>
    </ul>
  </div>
  <div
    className="css-xfzfsr"
  >
    <div
      className="css-10lc53e"
    >
      <div
        className="css-1gyptdh"
      >
        <p
          className="css-p619qv"
        >
          Annual Plan
        </p>
      </div>
      <div
        className="css-1qhzrdj"
      >
        <p
          className="css-p619qv"
        >
          $50.00
        </p>
      </div>
    </div>
    <div
      className="css-10lc53e"
    >
      <div
        className="css-1gyptdh"
      >
        <p
          className="css-p619qv"
        >
          Total
        </p>
      </div>
      <div
        className="css-aty8j4"
      >
        <p
          className="css-p619qv"
        >
          $50.00
        </p>
      </div>
    </div>
  </div>
  <form
    className="css-d92vw3"
    onSubmit={[Function]}
  >
    <div
      className="css-1wwdssh"
    >
      <label
        className="css-43qmoq"
        htmlFor="card-element"
      >
        Credit or Debit Card
      </label>
      <div
        className="css-1f994ut"
        id="card-element"
      />
      <div
        className="css-kqgi67"
      >
        <p
          className="css-1qyvk2s"
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-1mo5pof-MuiButtonBase-root-MuiButton-root"
      disabled={true}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "display": "flex",
          "justifySelf": "flex-end",
        }
      }
      tabIndex={-1}
      type="submit"
    >
      Subscribe
    </button>
  </form>
  <span
    className="css-1iupv5q"
  >
    <a
      href="https://stripe.com/"
    >
      <img
        alt="Stripe Badge"
        src="powered_by_stripe.png"
      />
    </a>
  </span>
</div>
`;
